modules/pwa/assets/js/service-worker.js

```diff
@@ -0,0 +1,271 @@
+/* jshint esversion: 6 */
+/* global self, caches, console, Promise, Request, pwa_vars_json */
+
+var CACHE = 'cache-v1';
+var pwa_vars = pwa_vars_json;
+var admin_regex = new RegExp( pwa_vars.admin_url );
+var site_regex = new RegExp( pwa_vars.site_url );
+
+// On install, cache some resources.
+self.addEventListener('install', function (evt) {
+    console.log('The service worker is being installed.');
+
+    // Ask the service worker to keep installing until the returning promise
+    // resolves.
+    evt.waitUntil( precache() );
+});
+
+self.addEventListener('activate', function(event) {
+    console.log('Service Worker activating.');
+
+	// https://developers.google.com/web/updates/2017/02/navigation-preload
+	if (self.registration.navigationPreload) {
+		event.waitUntil( self.registration.navigationPreload.enable() );
+	}
+
+	// Remove old caches
+    event.waitUntil(
+        caches.keys().then( function( cacheNames ) {
+            return Promise.all( cacheNames.map( function( key ) {
+                if( key !== CACHE ) {
+                    console.log( 'Service Worker: Removing Old Cache', key );
+                    return caches.delete( key );
+                }
+            } ) );
+        })
+	);
+
+    return self.clients.claim();
+});
+
+// On fetch, try the cache but if there's a miss try loading the content
+self.addEventListener('fetch', function (evt) {
+    if ( shouldCacheRequest( evt.request ) ) {
+		if ( isExternalAsset( evt.request.url ) ) {
+			evt.request.mode = 'no-cors';
+		}
+        evt.respondWith( fetchAndCache( evt.request, evt ) );
+    } else {
+        evt.respondWith( fetch( evt.request ) );
+    }
+});
+
+function isExternalAsset( url ) {
+	return ! site_regex.test( url ) && ! url.match(/^\/[^\/]/);
+}
+
+function fetchAndCache( request, event ) {
+    // open cache
+    return caches.open(CACHE).then( function( cache ) {
+		// find in cache
+
+		return cache.match( request )
+			.then( function( response ) {
+				// only do more work if we don't already have a response
+				if ( response ) {
```

Comment:
What's the invalidation process? If visitor caches content and site owner updates the resource.. Unlike the "regular" HTTP cache, resources in SW cache are not automatically checked for freshness and have to be manually (re+in)validated by our code here.

Reply:
I added stale-while-revalidate in the latest couple of commits. I'm not actually taking over the browser and replacing content yet, that's next...